drafts/when-elixirs-performance-becomes-rust-y.md

```diff
@@ -0,0 +1,405 @@
+        atom ok;
+        atom error;
+    }
+}
+```
+
+The first part remains the same, we'll declare the ok and error atoms for use with Elixir.
+
+```Rust
+rustler::rustler_export_nifs! {
+    "Elixir.RustNif.PrimeNumbers",
+    [
+        ("prime_numbers", 1, prime_numbers)
+    ],
+    None
+}
+```
+
+We'll register a `prime_numbers` method to rustler with an arity of 1 to be exported. This will be made available to Elixir.
+
+```Rust
+fn is_prime_number(x: i64) -> bool {
+    let end_of_range = x - 1;
+
+    if x == 1 {
+        false
+    } else {
+        !(2..end_of_range).any(|num| x % num == 0)
+    }
+}
+```
+
+We'll define a method to determine whether a number is a prime number. It accepts the number as argument and returns a boolean.
+
+```Rust
+fn prime_numbers<'a>(env: Env<'a>, args: &[Term<'a>]) -> Result<Term<'a>, Error> {
+    let is_list: bool = args[0].is_list();
+
+    if !is_list {
+        Ok((atoms::error(), "No list supplied").encode(env))
+    } else {
+        let numbers: Vec<i64> = args[0].decode()?;
+        let mut vec = Vec::new();
+
+        for number in numbers {
+            if is_prime_number(number) {
+                vec.push(number)
+            }
+        }
+
+        Ok((atoms::ok(), &*vec).encode(env))
+    }
+}
+```
+
+We can use the same function signature that the example add function used (note the <'a> are [lifetimes](https://doc.rust-lang.org/book/ch10-03-lifetime-syntax.html), not generics or similar).
+
+In this case we will verify whether the first argument is a list, and stop the function and yield `{:error, "No list suppled"}` if it's not.
+
+Afterwards, we use rustler's `decode()` method to convert the data type into a Vec of `i64`. We'll make a mutable Vec (you need to specify mutability explicitely in Rust), and loop through the numbers to determine the prime number. If a number is determined to be a prime number, we push it into the Vec.
+
+### Glueing Elixir and Rust together
+
+We'll need a module in Elixir that will be the interface to the Rust code. The module name corresponds to the name known in the Rust code.
+
+```Elixir
+# prime_numbers.ex
+defmodule RustNif.PrimeNumbers do
+  use Rustler, otp_app: :rust_nif, crate: :primenumbers
+
+  # Overriden when module is loaded
+  def prime_numbers(_nums), do: :erlang.nif_error(:nif_not_loaded)
+end
+```
+
+We define the OTP app we use, and the Rust crate we're loading, and that's it.
+
+## Comparison of performance
+
+Now that we have both the Elixir and Rust(Elixir) versions of the code, let's compare them on their execution speed!
+
+```Elixir
+defmodule Benchmark do
+  def benchmark do
+    Process.spawn(fn -> benchmark_Elixir() end, [:link])
+    Process.spawn(fn -> benchmark_rust() end, [:link])
+    nil
+  end
+
+  def benchmark_Elixir do
+    time = Time.utc_now
+    RustNif.ElixirPrimes.prime_numbers(Enum.into 1..100000, [])
+    |> IO.inspect
```

Comment:
The benchmark is a bit scewed as it now also includes printing the result :thinking:

Reply:
Right, I didn't think of that. I'll change it and re-run the benchmarks!